### PR TITLE
Parallax Corrected Cubemaps

### PR DESCRIPTION
**Pull request by BenLubar**
`comp_cubemap_parallax` is an entity that looks for nearby cubemaps and rewrites the materials generated by them to have a precomputed matrix used by the shader. (See #230.)

The code works:

However, I don't write a lot of Python, so it's possible I did something bad.

It doesn't rotate the bounding box if you rotate the entity in Hammer, but `comp_propcombine_set` has the same issue so I'm assuming there's just no good way to implement that feature in an FGD.

Rotating the bounding box gives you a somewhat strange cubemap because cubemaps are generated in axis-aligned space, but the corners match up with a room rotated the same amount, so it's working as well as it can be.

Other AS:RD FGD support is going to take a while as I have to go through 7+ years of entity changes, but I'm planning to just add `comp_cubemap_parallax` and `comp_propcombine_set` to the FGD we ship in the meantime.

I haven't made editor/comp_cubemap_parallax.vmt or editor/comp_cubemap_parallax.vtf yet.

### CODE REVIEW

**Comment by TeamSpen210** [comment]
In regards to the rotation, I'd expect a better way to use them is to have the room in an instance, let the collapsing modify the angles. Python code looks quite reasonable, though instead of using a regex on the material you can use [`srctools.vmt.Material`](https://srctools.readthedocs.io/en/latest/modules/vmt.html#srctools.vmt.Material) to parse, edit and resave in a more convenient fashion. Might also be better to first parse all the parallax ents, then modify each material so you can do things like match to the closest entity if multiple overlap.

**Comment by TeamSpen210** [comment]
For robustness, might also want to check how it behaves with an envmapped water brush (since VBSP sometimes patches a `$waterlevel` parameter in), and also with `WorldVertexTransition` on brushes (which VBSP converts to `LightmappedGeneric` automatically.) Could produce nested patch materials or filenames that confuse.

**Comment by BenLubar** [comment]
image
Looks like envmapped water already creates a nested patch, but the one the regex doesn't pick up doesn't need to be edited.

(Although, the water shader doesn't do parallax-corrected cubemaps, so it'll just be a few extra lines of VMT that don't do anything in this case.)

image
Also looks like the wvt_patch material just gets patched for the cubemap like any other material.

File: transforms/comp_cubemap_parallax.py [hunk 1]
```diff
@@ -122,14 +122,14 @@ def helper(a, b, x, y):
             )
             continue
 
-        config['used'] += 1
+        best_match['used'] += 1
```
Comment (BenLubar) [inline transforms/comp_cubemap_parallax.py:129]: oh whoops

File: transforms/comp_cubemap_parallax.py [hunk 2]
```diff
@@ -28,12 +28,7 @@ def comp_cubemap_parallax(ctx: Context):
         diff = maxs - mins
 
         # ensure bounding box has volume
-        if diff[0] == 0.0:
-            diff[0] = 1.0
-        if diff[1] == 0.0:
-            diff[1] = 1.0
-        if diff[2] == 0.0:
-            diff[2] = 1.0
+        diff.max((1.0, 1.0, 1.0))
```
Comment (BenLubar) [inline transforms/comp_cubemap_parallax.py:31]: This is technically different if the bounds are specified in the wrong order, but making it look wrong is probably enough of a hint to the mapper that they messed up their numbers.

Comment (TeamSpen210) [reply transforms/comp_cubemap_parallax.py:31]: Ah true, `<` not `==`.

**Review by TeamSpen210 — APPROVED** [review]
If you don't have anything else to add I'll merge this.

**Comment by BenLubar** [comment]
LGTM

**Pull request merged**